Move recent additions to DIBuilder to LLVMExt

bcardiff opened the pull request:
#8538 cause some issues in the packaging that this PR should fix.

As stated by https://github.com/crystal-lang/crystal/pull/8538#issuecomment-580254461 the C API used is available since llvm 7.0.

Instead of just porting the code to work with older LLVM I added an explicit call to the C API. This way I think is easier to identify what will be the benefits of a cleanup.

No other method in llvm_ext.cc does this today. But is a way to be more ready to clean up llvm_ext eventually.

@skuznetsov can you point me to what small program should stress these funciontions / definitions?

skuznetsov commented:
@bcardiff Can you elaborate more what are you expecting from me as it is not really clear which functions / definitions you want me to script the test? The ones you've changed?

bcardiff commented:
@skuznetsov I wanted to know a crystal program that will end up using `create_lexical_block_file` and `create_unspecified_type` functions in the debug information that will show up in a lldb session somehow. 

This PR make things build and the very simple specs in ./spec/debug work for llvm 6, 7, 10. But that's all that I can check so far.

ysbaddaden commented:
These functions are mere polyfills. There is no need for wrapper on LLVM 7.0 and no compatibility issues. Why not just _implement_ them for LLVM < 7.0?

```cpp

#if LLVM_VERSION_LT(7, 0)
LLVMMetadataRef LLVMDIBuilderCreateUnspecifiedType(DIBuilderRef Dref, const char *Name, size_t NameLen) {
  return wrap(Dref->createUnspecifiedType({Name, NameLen}));
}

LLVMMetadataRef LLVMDIBuilderCreateLexicalBlockFile(DIBuilderRef Dref, LLVMMetadataRef Scope, LLVMMetadataRef File, unsigned Discriminator) {
  return wrap(Dref->createLexicalBlockFile(unwrapDI<DIScope>(Scope), unwrapDI<DIFile>(File), Discriminator));
}
#endif
```

oprypin commented:
Mm I had a different issue with this myself, was confused why my approach stopped working suddenly.
This may be an "opposite" issue, because this is with *LLVM 10*.

> LLVMCore.lib(DebugInfo.obj) : error LNK2005: LLVMDIBuilderCreateLexicalBlockFile already defined in LLVM-C.lib(LLVM-C.dll)
> LLVMCore.lib(DebugInfo.obj) : error LNK2005: LLVMDIBuilderCreateUnspecifiedType already defined in LLVM-C.lib(LLVM-C.dll)

https://github.com/oprypin/crystal/runs/597746581#step:18:114

My workaround was to delete "llvm\lib\LLVM-C.lib" before being able to link "src\llvm\ext\llvm_ext.obj" "llvm\Release\lib\LLVM*.lib"

Note that this happens both with and without this pull request.

oprypin commented:
I disagree with @ysbaddaden suggestion, which would basically invade on LLVM's namespaces and likely cause issues. What if you compiled against an older LLVM but then linked against a newer one? As I understand, that'll just break.

oprypin commented on src/llvm/ext/llvm_ext.cc:
```diff
@@ -272,6 +272,27 @@ LLVMMetadataRef LLVMExtDIBuilderCreateReplaceableCompositeType(
   return wrap(CT);
 }
 
+LLVMMetadataRef LLVMExtDIBuilderCreateUnspecifiedType(
+  DIBuilderRef Dref, const char *Name, size_t NameLen) {
+#if LLVM_VERSION_GE(7, 0)
+  return LLVMDIBuilderCreateUnspecifiedType(wrap(Dref), Name, NameLen);
+#else
+  return wrap(Dref->createUnspecifiedType({Name, NameLen}));
+#endif
+}
+
+LLVMMetadataRef LLVMExtDIBuilderCreateLexicalBlockFile(
+  DIBuilderRef Dref,
+  LLVMMetadataRef Scope, LLVMMetadataRef File, unsigned Discriminator) {
+#if LLVM_VERSION_GE(7, 0)
+  return LLVMDIBuilderCreateLexicalBlockFile(wrap(Dref), Scope, File, Discriminator);
+#else
+  return wrap(Dref->createLexicalBlockFile(unwrapDI<DIScope>(Scope),
+                                           unwrapDI<DIFile>(File),
+                                           Discriminator));
+#endif
+}
```
```suggestion
LLVMMetadataRef LLVMExtDIBuilderCreateUnspecifiedType(
  DIBuilderRef Dref, const char *Name, size_t NameLen) {
  return wrap(Dref->createUnspecifiedType({Name, NameLen}));
}

LLVMMetadataRef LLVMExtDIBuilderCreateLexicalBlockFile(
  DIBuilderRef Dref,
  LLVMMetadataRef Scope, LLVMMetadataRef File, unsigned Discriminator) {
  return wrap(Dref->createLexicalBlockFile(unwrapDI<DIScope>(Scope),
                                           unwrapDI<DIFile>(File),
                                           Discriminator));
}
```

I think it's best to just implement the function directly, without conditionals, and add a TODO to use LLVM instead when older support is dropped.

This is the only thing that helped in my case.
https://github.com/oprypin/crystal/runs/597786291#step:18:109

The main thing that might be causing problems for me is that this introduces a dependency on LLVM-C in llvm_ext?

skuznetsov commented:
> @skuznetsov I wanted to know a crystal program that will end up using `create_lexical_block_file` and `create_unspecified_type` functions in the debug information that will show up in a lldb session somehow.
`create_lexical_block_file` was the only solution for blocks (`each`, `loop`, etc) when you have to incorporate references to different file because `create_lexical_block` works only on current scope file.

So example:
```
[1,2,3].each do |val|
   puts val
end
```
should generate that call for you

`create_unspecified_type` is used to create a true Nil class that not part of any union.

```
nil_value = Nil
puts nil_value
```
should call unspecified type.
> 
> This PR make things build and the very simple specs in ./spec/debug work for llvm 6, 7, 10. But that's all that I can check so far.

bcardiff commented:
I removed the wrap and left a comment. Simpler.

@skuznetsov I added a block example. But I couldn't find a way to show in lldb the type name for `decltype(nullptr)`  🤔

oprypin commented:
CI fails with
```
src/llvm/ext/llvm_ext.cc:1:10: fatal error: 'llvm/IR/DIBuilder.h' file not found
#include "llvm/IR/DIBuilder.h"
         ^~~~~~~~~~~~~~~~~~~~~
```
How is that possible? Nothing has changed in that regard.
Maybe it was already to fail but now we tripped some cache?

bcardiff commented:
> /usr/local/Cellar/llvm/10.0.0_1: 7,019 files, 1GB in https://circleci.com/gh/crystal-lang/crystal/42980
vs
> /usr/local/Cellar/llvm/10.0.0_2: 7,055 files, 1GB in https://circleci.com/gh/crystal-lang/crystal/43132

in `bin/ci prepare_build` maybe 😞

skuznetsov commented:
> @skuznetsov I added a block example. But I couldn't find a way to show in lldb the type name for `decltype(nullptr)` 🤔

This is interesting. The only call that generates `!DIBasicType(tag: DW_TAG_unspecified_type, name: "decltype(nullptr)")` is not my example:

```
val = Nil
puts val
```

but code in src/string.cr:
```
  # :nodoc:
  def self.check_capacity_in_bounds(capacity)
    if capacity < 0
      raise ArgumentError.new("Negative capacity")
    end

    if capacity.to_u64 > (UInt32::MAX - HEADER_SIZE - 1)
      raise ArgumentError.new("Capacity too big")
    end
  end
```
in this case this is static method and self is Nil.

Here is what was generated in IR file:
```
!987 = distinct !DISubprogram(name: "check_capacity_in_bounds", linkageName: "check_capacity_in_bounds", scope: !589, file: !589, line: 4715, type: !988, scopeLine: 4715, spFlags: DISPFlagLocalToUnit | DI
!988 = !DISubroutineType(types: !989)
!989 = !{!990, !4}
!990 = !DIBasicType(tag: DW_TAG_unspecified_type, name: "decltype(nullptr)")
!991 = !{!992}
!992 = !DILocalVariable(name: "capacity", arg: 1, scope: !987, file: !589, line: 4715, type: !4)
!993 = !DILocation(line: 4715, column: 3, scope: !987)
!994 = !DILocation(line: 4716, column: 5, scope: !987)
!995 = !DILocation(line: 4717, column: 13, scope: !987)
!996 = !DILocation(line: 4717, column: 7, scope: !987)
!997 = !DILocation(line: 4720, column: 5, scope: !987)
!998 = !DILocation(line: 4721, column: 13, scope: !987)
!999 = !DILocation(line: 4721, column: 7, scope: !987)
!1000 = !DILocation(line: 4720, column: 39, scope: !987)
!1001 = !DILocation(line: 4720, column: 53, scope: !987)
```

skuznetsov commented on spec/debug/blocks.cr:
```diff
@@ -0,0 +1,12 @@
+def f(v)
+  yield(v)
+end
+
+f("hello world") do |v|
```
I recommend to replace it with 
```
["hello world"].each do |v|
...
```
as it will generate proper testing as in this IR debug output:
```
!549 = distinct !DILexicalBlock(scope: !542, file: !550, line: 1, column: 1)
!550 = !DIFile(filename: "blocks.cr", directory: "/Users/sergey/Projects/crystal/crystal/test")
!551 = !DIDerivedType(tag: DW_TAG_pointer_type, name: "String", baseType: !552, size: 64, align: 64)
!552 = !DICompositeType(tag: DW_TAG_structure_type, name: "String", line: 1, size: 128, align: 128, elements: !553)
!553 = !{!554, !555, !556}
!554 = !DIDerivedType(tag: DW_TAG_member, name: "bytesize", line: 1, baseType: !4, size: 32, align: 32, offset: 32)
!555 = !DIDerivedType(tag: DW_TAG_member, name: "length", line: 1, baseType: !4, size: 32, align: 32, offset: 64)
!556 = !DIDerivedType(tag: DW_TAG_member, name: "c", line: 1, baseType: !120, size: 8, align: 8, offset: 96)
!557 = !DILocalVariable(name: "a", scope: !549, file: !550, line: 6, type: !551)
!558 = !DILocalVariable(name: "b", scope: !549, file: !550, line: 7, type: !4, align: 32)
!559 = !DILocalVariable(name: "v", scope: !549, file: !550, line: 1, type: !551)
!560 = !DILocation(line: 83, column: 40, scope: !561)
!561 = distinct !DILexicalBlock(scope: !542, file: !562, line: 1, column: 1)
!562 = !DIFile(filename: "hasher.cr", directory: "/Users/sergey/Projects/crystal/crystal/src/crystal")
!563 = !DILocation(line: 42, column: 1, scope: !564)
!564 = distinct !DILexicalBlock(scope: !542, file: !565, line: 1, column: 1)
!565 = !DIFile(filename: "kernel.cr", directory: "/Users/sergey/Projects/crystal/crystal/src")
!566 = !DILocation(line: 27, column: 23, scope: !567)
!567 = distinct !DILexicalBlock(scope: !542, file: !568, line: 1, column: 1)
!568 = !DIFile(filename: "lib_pcre.cr", directory: "/Users/sergey/Projects/crystal/crystal/src/regex")
!569 = !DILocation(line: 28, column: 21, scope: !567)
```

bcardiff commented:
Rebased on master due to #9129.
Squashed some changes and use `Array#each` for block example.

The scope is enough to get nightlies back.

skuznetsov commented on spec/debug/blocks.cr:
```diff
@@ -0,0 +1,12 @@
+def f(v)
+  yield(v)
+end
+
```
This code is useless now. You can remove it as it do nothing due to `Array#each`
